Surface_mesh_topology/include/CGAL/Surface_mesh_topology/internal/Shortest_noncontractible_cycle.h

```diff
@@ -0,0 +1,603 @@
+    // Add to queue the degree-1 faces
+    for (auto it=get_local_map().darts().begin(), itend=get_local_map().darts().end();
+         it!=itend; ++it)
+    {
+      if (get_local_map().is_marked(it, m_is_perforated))
+      {
+        if (!get_local_map().is_marked(it, face_deleted))
+        { get_local_map().template mark_cell<2>(it, face_deleted); }
+      }
+      else if (!get_local_map().is_marked(it, tested))
+      {
+        Dart_handle dh=it, dh_only_edge=nullptr;
+        bool degree_one=true;
+        do // Here we do not use is_degree_one_face method because we want to
+        {  // mark tested all darts of the face in the same loop.
+          get_local_map().template mark_cell<1,1>(dh, tested); // For CMap and GMap
+          if (degree_one)
+          {
+            if (!get_local_map().is_marked(dh, edge_deleted))
+            {
+              if (dh_only_edge!=nullptr) { degree_one=false; }
+              else { dh_only_edge=dh; }
+            }
+          }
+          dh=get_local_map().next(dh);
+        }
+        while(dh!=it);
+
+        if (degree_one && dh_only_edge!=nullptr)
+        {
+          degree_one_faces.push(dh_only_edge);
+          CGAL_assertion(!get_local_map().is_marked(dh_only_edge, edge_deleted));
+          CGAL_assertion(!get_local_map().is_marked(dh_only_edge, face_deleted));
+        }
+      }
+    }
+
+    // Remove the degree-1 faces
+    while (!degree_one_faces.empty())
+    {
+      Dart_handle dh_face=degree_one_faces.front();
+      degree_one_faces.pop();
+      if (!get_local_map().is_marked(dh_face, edge_deleted))
+      { get_local_map().template mark_cell<1>(dh_face, edge_deleted); }
+      if (!get_local_map().is_marked(dh_face, face_deleted))
+      { get_local_map().template mark_cell<2>(dh_face, face_deleted); }
+      Dart_handle dh_adj_face=get_local_map().opposite2(dh_face);
+      if (!get_local_map().is_marked(dh_adj_face, face_deleted))
+      {
+        Dart_handle dh_only_edge=nullptr;
+        if (is_degree_one_face(dh_adj_face, dh_only_edge, edge_deleted))
+        {
+          degree_one_faces.push(dh_only_edge);
+          CGAL_assertion(!get_local_map().is_marked(dh_only_edge, edge_deleted));
+          CGAL_assertion(!get_local_map().is_marked(dh_only_edge, face_deleted));
+        }
+      }
+    }
+    for (auto it=get_local_map().darts().begin(),
+           itend=get_local_map().darts().end(); it!=itend; ++it)
+    {
+      if (it<get_local_map().opposite2(it) && vertex_info(it)>=0 &&
+          !get_local_map().is_marked(it, edge_deleted))
+      { noncon_edges.push_back(it); }
+
+      get_local_map().unmark(it, tested);
+      get_local_map().unmark(it, edge_deleted);
+      get_local_map().unmark(it, face_deleted);
+    }
+
+    get_local_map().free_mark(tested);
+    get_local_map().free_mark(edge_deleted);
+    get_local_map().free_mark(face_deleted);
+  }
+
+  Dart_handle nonhole_dart_of_same_edge(Dart_handle dh)
```

Comment:
non_hole?

Reply:
Undocumented internal method; thus I suppose not really important.

Reply:
undocumented internal method; not really important.